File: mrnes/trace.go
package mrnes

import (
	"encoding/json"
	"github.com/iti/evt/vrtime"
	"gopkg.in/yaml.v3"
	"os"
	"path"
	"sort"
	"strconv"
)

type TraceRecordType int

const (
	NetworkType TraceRecordType = iota
	CmpPtnType
)

type TraceInst struct {
	TraceTime string
	TraceType string
	TraceStr  string
}

// NameType is a an entry in a dictionary created for a trace
// that maps object id numbers to a (name,type) pair
type NameType struct {
	Name string
	Type string
}

// TraceManager implements the pces TraceManager interface. It is
// use to gather information about a simulation model and an execution of that model
type TraceManager struct {
	// experiment uses trace
	InUse bool `json:"inuse" yaml:"inuse"`

	// name of experiment
	ExpName string `json:"expname" yaml:"expname"`

	// text name associated with each objID
	NameByID map[int]NameType `json:"namebyid" yaml:"namebyid"`

	// all trace records for this experiment
	Traces map[int][]TraceInst `json:"traces" yaml:"traces"`
}

// CreateTraceManager is a constructor.  It saves the name of the experiment
// and a flag indicating whether the trace manager is active.  By testing this
// flag we can inhibit the activity of gathering a trace when we don't want it,
// while embedding calls to its methods everywhere we need them when it is
func CreateTraceManager(ExpName string, active bool) *TraceManager {
	tm := new(TraceManager)
	tm.InUse = active
	tm.ExpName = ExpName
	tm.NameByID = make(map[int]NameType)  // dictionary of id code -> (name,type)
	tm.Traces = make(map[int][]TraceInst) // traces have 'execution' origins, are saved by index to these
	return tm
}

// Active tells the caller whether the Trace Manager is actively being used
func (tm *TraceManager) Active() bool {
	return tm.InUse
}

// AddTrace creates a record of the trace using its calling arguments, and stores it
func (tm *TraceManager) AddTrace(vrt vrtime.Time, execID int, trace TraceInst) {

	// return if we aren't using the trace manager
	if !tm.InUse {
		return
	}

	_, present := tm.Traces[execID]
	if !present {
		tm.Traces[execID] = make([]TraceInst, 0)
	}
	tm.Traces[execID] = append(tm.Traces[execID], trace)
}

// AddName is used to add an element to the id -> (name,type) dictionary for the trace file
func (tm *TraceManager) AddName(id int, name string, objDesc string) {
	if tm.InUse {
		_, present := tm.NameByID[id]
		if present {
			panic("duplicated id in AddName")
		}
		tm.NameByID[id] = NameType{Name: name, Type: objDesc}
	}
}

// WriteToFile stores the Traces struct to the file whose name is given.
// Serialization to json or to yaml is selected based on the extension of this name.
func (tm *TraceManager) WriteToFile(filename string, globalOrder bool) bool {
	if !tm.InUse {
		return false
	}
	pathExt := path.Ext(filename)
	var bytes []byte
	var merr error = nil

	if !globalOrder {
		if pathExt == ".yaml" || pathExt == ".YAML" || pathExt == ".yml" {
			bytes, merr = yaml.Marshal(*tm)
		} else if pathExt == ".json" || pathExt == ".JSON" {
			bytes, merr = json.MarshalIndent(*tm, "", "\t")
		}

		if merr != nil {
			panic(merr)
		}
	} else {
		ntm := new(TraceManager)
		ntm.InUse = tm.InUse
		ntm.ExpName = tm.ExpName
		ntm.NameByID = make(map[int]NameType)
		for key, value := range tm.NameByID {
			ntm.NameByID[key] = value
		}
		ntm.Traces = make(map[int][]TraceInst)
		ntm.Traces[0] = make([]TraceInst, 0)
		for _, valueList := range tm.Traces {
			ntm.Traces[0] = append(ntm.Traces[0], valueList...)
		}

		sort.Slice(ntm.Traces[0], func(i, j int) bool {
			v1, _ := strconv.ParseFloat(ntm.Traces[0][i].TraceTime, 64)
			v2, _ := strconv.ParseFloat(ntm.Traces[0][j].TraceTime, 64)
			return v1 < v2
		})
		if pathExt == ".yaml" || pathExt == ".YAML" || pathExt == ".yml" {
			bytes, merr = yaml.Marshal(*ntm)
		} else if pathExt == ".json" || pathExt == ".JSON" {
			bytes, merr = json.MarshalIndent(*ntm, "", "\t")
		}

		if merr != nil {
			panic(merr)
		}
	}

	f, cerr := os.Create(filename)
	if cerr != nil {
		panic(cerr)
	}
	_, werr := f.WriteString(string(bytes[:]))
	if werr != nil {
		panic(werr)
	}
	err := f.Close()
	if err != nil {
		panic(err)
	}
	return true
}

// NetTrace saves information about the visitation of a message to some point in the simulation.
// saved for post-run analysis
type NetTrace struct {
	Time      float64 // time in float64
	Ticks     int64   // ticks variable of time
	Priority  int64   // priority field of time-stamp
	FlowID    int     // integer identifier identifying the chain of traces this is part of
	ExecID    int
	ConnectID int    // integer identifier of the network connection
	ObjID     int    // integer id for object being referenced
	Op        string // "start", "stop", "enter", "exit"
	PcktIdx   int    // packet index inside of a multi-packet message
	Packet    bool   // true if the event marks the passage of a packet (rather than flow)
	MsgType   string
	Rate      float64 // rate associated with the connection
}

func (ntr *NetTrace) TraceType() TraceRecordType {
	return NetworkType
}

func (ntr *NetTrace) Serialize() string {
	var bytes []byte
	var merr error

	bytes, merr = yaml.Marshal(*ntr)

	if merr != nil {
		panic(merr)
	}
	return string(bytes[:])
}

type SchedulerTrace struct {
	Time      float64
	ObjID     int
	ExecID    int
	Op        string
	Cores     int
	Inservice int
	Inbckgrnd int
	Waiting   int
}

func (st *SchedulerTrace) Serialize() string {
	var bytes []byte
	var merr error

	bytes, merr = yaml.Marshal(*st)

	if merr != nil {
		panic(merr)
	}
	return string(bytes[:])
}

type IntrfcTrace struct {
	Time   float64
	ObjID  int
	ExecID int
	MsgID  int
	Op     string
	CQ     string
}

func (it *IntrfcTrace) Serialize() string {
	var bytes []byte
	var merr error

	bytes, merr = yaml.Marshal(*it)

	if merr != nil {
		panic(merr)
	}
	return string(bytes[:])
}

func AddIntrfcTrace(tm *TraceManager, vrt vrtime.Time, execID, msgID, objID int, op, CQ string) {
	it := new(IntrfcTrace)
	it.Time = vrt.Seconds()
	it.ExecID = execID
	it.MsgID = msgID
	it.ObjID = objID
	it.Op = op
	it.CQ = CQ
	itStr := it.Serialize()
	strTime := strconv.FormatFloat(it.Time, 'f', -1, 64)
	trcInst := TraceInst{TraceTime: strTime, TraceType: "interface", TraceStr: itStr}
	tm.AddTrace(vrt, execID, trcInst)
}

func AddSchedulerTrace(tm *TraceManager, vrt vrtime.Time, ts *TaskScheduler, execID, objID int, op string) {
	endpt := EndptDevByID[objID]
	if !endpt.EndptState.Trace {
		return
	}

	st := new(SchedulerTrace)
	st.Time = vrt.Seconds()
	st.ExecID = execID
	st.ObjID = objID
	st.Op = op
	st.Cores = ts.cores
	st.Inservice = ts.inservice
	st.Waiting = ts.numWaiting
	st.Inbckgrnd = ts.inBckgrnd
	stStr := st.Serialize()

	traceTime := strconv.FormatFloat(vrt.Seconds(), 'f', -1, 64)
	trcInst := TraceInst{TraceTime: traceTime, TraceType: "scheduler", TraceStr: stStr}
	tm.AddTrace(vrt, st.ExecID, trcInst)
}

// AddNetTrace creates a record of the trace using its calling arguments, and stores it
func AddNetTrace(tm *TraceManager, vrt vrtime.Time, nm *NetworkMsg, objID int, op string) {
	ntr := new(NetTrace)
	ntr.Time = vrt.Seconds()
	ntr.Ticks = vrt.Ticks()
	ntr.Priority = vrt.Pri()
	ntr.ConnectID = nm.ConnectID
	ntr.FlowID = nm.FlowID
	ntr.ObjID = objID
	ntr.ExecID = nm.ExecID
	ntr.Op = op
	ntr.PcktIdx = nm.PcktIdx
	ntr.MsgType = nmtToStr[nm.NetMsgType]

	ntrStr := ntr.Serialize()
	traceTime := strconv.FormatFloat(vrt.Seconds(), 'f', -1, 64)

	trcInst := TraceInst{TraceTime: traceTime, TraceType: "network", TraceStr: ntrStr}
	tm.AddTrace(vrt, ntr.ExecID, trcInst)
}

File: mrnes/trace.py
"""
Python translation of mrnes/trace.go
Implements TraceManager and related trace record types for simulation tracing and serialization to JSON/YAML.

This module is a direct translation of Go's trace.go, including all comments and documentation.
It provides tracing infrastructure for simulation models, including trace records, trace manager, and serialization to JSON/YAML.
"""
import json
import yaml
import pathlib
from typing import Dict, List, Any
import evt_python.evt.vrtime as vrtime

class TraceRecordType:
    # TraceRecordType is an enum for trace record types
    NetworkType = 0  # NetworkType TraceRecordType = iota
    CmpPtnType = 1

class TraceInst:
    """
    TraceInst holds a single trace record, including time, type, and serialized string.
    """
    def __init__(self, TraceTime: str, TraceType: str, TraceStr: str):
        self.TraceTime = TraceTime
        self.TraceType = TraceType
        self.TraceStr = TraceStr

class NameType:
    """
    NameType is an entry in a dictionary created for a trace
    that maps object id numbers to a (name,type) pair
    """
    def __init__(self, Name: str, Type: str):
        self.Name = Name
        self.Type = Type

class TraceManager:
    """
    TraceManager implements the pces TraceManager interface. It is
    used to gather information about a simulation model and an execution of that model.
    """
    def __init__(self, ExpName: str, active: bool):
        # experiment uses trace
        self.InUse = active
        # name of experiment
        self.ExpName = ExpName
        # text name associated with each objID
        self.NameByID: Dict[int, NameType] = {}
        # all trace records for this experiment
        self.Traces: Dict[int, List[TraceInst]] = {}

    def active(self) -> bool:
        """
        Active tells the caller whether the Trace Manager is actively being used
        """
        return self.InUse

    def add_trace(self, vrt: Any, execID: int, trace: TraceInst):
        """
        AddTrace creates a record of the trace using its calling arguments, and stores it
        """
        # return if we aren't using the trace manager
        if not self.InUse:
            return
        if execID not in self.Traces:
            self.Traces[execID] = []
        self.Traces[execID].append(trace)

    def add_name(self, id: int, name: str, objDesc: str):
        """
        AddName is used to add an element to the id -> (name,type) dictionary for the trace file
        """
        if self.InUse:
            if id in self.NameByID:
                raise Exception("duplicated id in AddName")
            self.NameByID[id] = NameType(name, objDesc)

    def write_to_file(self, filename: str, globalOrder: bool) -> bool:
        """
        WriteToFile stores the Traces struct to the file whose name is given.
        Serialization to json or to yaml is selected based on the extension of this name.
        """
        if not self.InUse:
            return False
        
        path_ext = pathlib.Path(filename).suffix.lower()
        if not globalOrder:
            data = self._serialize()
        else:
            data = self._serialize_global_order()
        
        if path_ext in ['.yaml', '.yml']:
            bytestr = yaml.dump(data, sort_keys=False)
        elif path_ext == '.json':
            bytestr = json.dumps(data, indent=2)
        else:
            raise Exception(f"Unsupported file extension: {path_ext}")
        
        with open(filename, 'w') as f:
            f.write(bytestr)

        return True

    def _serialize(self):
        return {
            'inuse': self.InUse,
            'expname': self.ExpName,
            'namebyid': {k: vars(v) for k, v in self.NameByID.items()},
            'traces': {k: [vars(t) for t in v] for k, v in self.Traces.items()}
        }

    def _serialize_global_order(self):
        ntm = {
            'inuse': self.InUse,
            'expname': self.ExpName,
            'namebyid': {k: vars(v) for k, v in self.NameByID.items()},
            'traces': {}
        }
        all_traces = []
        for value_list in self.Traces.values():
            all_traces.extend(value_list)
        all_traces.sort(key=lambda t: float(t.TraceTime))
        ntm['traces'][0] = [vars(t) for t in all_traces]
        return ntm

class NetTrace:
    """
    NetTrace saves information about the visitation of a message to some point in the simulation.
    Saved for post-run analysis.
    """
    def __init__(self, Time, Ticks, Priority, FlowID, ExecID, ConnectID, ObjID, Op, PcktIdx, Packet, MsgType, Rate):
        self.Time = Time  # time in float64
        self.Ticks = Ticks  # ticks variable of time
        self.Priority = Priority  # priority field of time-stamp
        self.FlowID = FlowID  # integer identifier identifying the chain of traces this is part of
        self.ExecID = ExecID
        self.ConnectID = ConnectID  # integer identifier of the network connection
        self.ObjID = ObjID  # integer id for object being referenced
        self.Op = Op  # "start", "stop", "enter", "exit"
        self.PcktIdx = PcktIdx  # packet index inside of a multi-packet message
        self.Packet = Packet  # true if the event marks the passage of a packet (rather than flow)
        self.MsgType = MsgType
        self.Rate = Rate  # rate associated with the connection
    
    def trace_type(self):
        return TraceRecordType.NetworkType
    
    def serialize(self):
        return yaml.dump(vars(self), sort_keys=False)

class SchedulerTrace:
    """
    SchedulerTrace saves information about scheduling events in the simulation.
    """
    def __init__(self, Time, ObjID, ExecID, Op, Cores, Inservice, Inbckgrnd, Waiting):
        self.Time = Time
        self.ObjID = ObjID
        self.ExecID = ExecID
        self.Op = Op
        self.Cores = Cores
        self.Inservice = Inservice
        self.Inbckgrnd = Inbckgrnd
        self.Waiting = Waiting
    
    def serialize(self):
        return yaml.dump(vars(self), sort_keys=False)

class IntrfcTrace:
    """
    IntrfcTrace saves information about interface events in the simulation.
    """
    def __init__(self, Time, ObjID, ExecID, MsgID, Op, CQ):
        self.Time = Time
        self.ObjID = ObjID
        self.ExecID = ExecID
        self.MsgID = MsgID
        self.Op = Op
        self.CQ = CQ
    
    def serialize(self):
        return yaml.dump(vars(self), sort_keys=False)

def add_intrfc_trace(tm: TraceManager, vrt: vrtime.Time, execID, msgID, objID, op, CQ):
    """
    AddIntrfcTrace creates a record of the trace using its calling arguments, and stores it
    """
    it = IntrfcTrace(Time=vrt.Seconds(), ObjID=objID, ExecID=execID, MsgID=msgID, Op=op, CQ=CQ)
    it_str = it.serialize()
    str_time = str(it.Time) # TODO: make sure this prints the same as Go version
    trc_inst = TraceInst(TraceTime=str_time, TraceType="interface", TraceStr=it_str)
    tm.add_trace(vrt, execID, trc_inst)

def add_scheduler_trace(tm: TraceManager, vrt: vrtime.Time, ts, execID, objID, op):
    """
    AddSchedulerTrace creates a record of the trace using its calling arguments, and stores it
    ts is expected to have .cores, .inservice, .numWaiting, .inBckgrnd attributes
    """
    endpt = EndptDevById[objID]
    if not endpt.EndptState.Trace:
        return
    
    st = SchedulerTrace(Time=vrt.Seconds(), ObjID=objID, ExecID=execID, Op=op,
                        Cores=ts.cores, Inservice=ts.inservice, Inbckgrnd=ts.inBckgrnd, Waiting=ts.numWaiting)
    st_str = st.serialize()

    trace_time = str(vrt.Seconds()) # TODO: make sure this prints the same as Go version
    trc_inst = TraceInst(TraceTime=trace_time, TraceType="scheduler", TraceStr=st_str)
    tm.add_trace(vrt, execID, trc_inst)

def add_net_trace(tm: TraceManager, vrt: vrtime.Time, nm, objID, op):
    """
    AddNetTrace creates a record of the trace using its calling arguments, and stores it
    nm is expected to have .ConnectID, .FlowID, .ExecID, .PcktIdx, .NetMsgType attributes
    """
    ntr = NetTrace(Time=vrt.Seconds(), Ticks=vrt.Ticks(), Priority=vrt.Pri(),
                   FlowID=nm.FlowID, ExecID=nm.ExecID, ConnectID=nm.ConnectID, ObjID=objID,
                   Op=op, PcktIdx=nm.PcktIdx, Packet=False,
                   MsgType=nmtToStr[nm.NetMsgType], Rate=0.0)
    
    ntr_str = ntr.serialize()
    trace_time = str(vrt.Seconds()) # TODO: make sure this prints the same as Go version
    
    trc_inst = TraceInst(TraceTime=trace_time, TraceType="network", TraceStr=ntr_str)
    tm.add_trace(vrt, ntr.ExecID, trc_inst)
